### PR TITLE
Fix more PodMutationHook issues for backwards compatibility

### PR DESCRIPTION
**Pull request by kaxil**
This PR/commit
- Adds missing affinity from old POD
- Adds comprehensive tests to check pod_mutation_hook works well with both new and old PODs with various configs like volume, volumeMounts, Ports, affinity, tolerations etc
- Refactors various parts of k8s code

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Review by kaxil — COMMENTED** [review]
cc @aneesh-joseph Thanks for testing the last fix, can you please help me by testing this PR too -- Can you run your pod_mutations hooks with this PR, please (both pre 1.10.11 and 1.10.11 behaviors) ?

**Comment by aneesh-joseph** [comment]
> cc @aneesh-joseph Thanks for testing the last fix, can you please help me by testing this PR too -- Can you run your pod_mutations hooks with this PR, please (both pre 1.10.11 and 1.10.11 behaviors) ?

sure, will test it out and slack you

**Comment by dimberman** [comment]
addresses https://github.com/apache/airflow/issues/9827 as well

**Comment by dimberman** [comment]
`│                                                                                      │
│ Traceback (most recent call last):                                                   │
│   File "/home/airflow/.local/bin/airflow", line 25, in <module>                      │
│     from airflow.configuration import conf                                           │
│   File "/home/airflow/.local/lib/python3.6/site-packages/airflow/__init__.py", line  │
│     from airflow.utils.log.logging_mixin import LoggingMixin                         │
│   File "/home/airflow/.local/lib/python3.6/site-packages/airflow/utils/__init__.py", │
│     from .decorators import apply_defaults as _apply_defaults                        │
│   File "/home/airflow/.local/lib/python3.6/site-packages/airflow/utils/decorators.py │
│     from airflow import settings                                                     │
│   File "/home/airflow/.local/lib/python3.6/site-packages/airflow/settings.py", line  │
│     from airflow.configuration import conf, AIRFLOW_HOME, WEBSERVER_CONFIG  # NOQA F │
│   File "/home/airflow/.local/lib/python3.6/site-packages/airflow/configuration.py",  │
│     with open(AIRFLOW_CONFIG, 'w') as f:                                             │
│ IsADirectoryError: [Errno 21] Is a directory: '/opt/airflow/airflow.cfg'             │
│ stream closed                                                                        │
│`

that is causing k8s tests to fail

**Comment by dimberman** [comment]
scheduler
```
│   config:                                                                            │
│     Type:      ConfigMap (a volume populated by a ConfigMap)                         │
│     Name:      airflow-airflow-config                                                │
│     Optional:  false
```

worker
```
   airflow-config:                                                                    │
│     Type:       EmptyDir (a temporary directory that shares a pod's lifetime)        │
│     Medium:                                                                          │
│     SizeLimit:  <unset>
```

for some reason the workers are mounting the cfg as an emptydir

**Comment by dimberman** [comment]
@kaxil so the tasks are completing correctly but then the pod goes into a crash loop and I'm not sure why. Will confer with you in the morning.

**Comment by dimberman** [comment]
@kaxil everything is working except those random mongo_hook tests. Can you take a look? After that we just need to refactor.

At the very least k8s stuff is WAY better tested now :)

**Comment by kaxil** [comment]
> @kaxil everything is working except those random mongo_hook tests. Can you take a look? After that we just need to refactor.
> 
> At the very least k8s stuff is WAY better tested now :)

Aneesh has reported that pod_mutation_hook isn't working for pre-1.10.11

File: airflow/kubernetes/pod_generator.py [hunk 1]
```diff
@@ -637,6 +640,37 @@ def extend_object_field(base_obj, client_obj, field_name):
...
+            client_obj_set[k] = merge_objects(v, client_obj_set[k])
+    appended_field_keys = sorted(client_obj_set.keys())
+    appended_fields = [client_obj_set[k] for k in appended_field_keys]
+    return appended_fields
+
+
+def get_dict_from_list(base_list):
+    """
+    :param base_list:
+    :type base_list: list(Optional[dict,
```
Comment (kaxil) [inline airflow/kubernetes/pod_generator.py]: We need to fix this or remove the type and make it an internal Method

File: airflow/kubernetes/pod_generator.py [hunk 2]
```diff
@@ -36,6 +36,9 @@
 import kubernetes.client.models as k8s
 import yaml
 from kubernetes.client.api_client import ApiClient
+from ..contrib.kubernetes.pod import (
```
Comment (turbaszek) [inline airflow/kubernetes/pod_generator.py]: Why not use full path?

Comment (dimberman) [reply airflow/kubernetes/pod_generator.py]: @turbaszek it seems you need to use relative paths if you want to use private functions

Comment (turbaszek) [reply airflow/kubernetes/pod_generator.py]: Hm, I checked out this code in both cases `..` and `airflow.contrib` I see warning `Access to a protected member _extract_volume_mounts of a module`. So I don't think this is a problem. If there's any particular reason to have such import I would be happy to have a comment on why we do this way

Comment (kaxil) [reply airflow/kubernetes/pod_generator.py]: Fixed

**Comment by kaxil** [comment]
A big thanks and a shoutout to @aneesh-joseph for helping us with testing that both pre-1.10.11 and 1.10.11 behaviours are maintained for 1.10.12. 🚀

**Comment by kaxil** [comment]
Please also refer to https://github.com/apache/airflow/commit/c230156739178762d5cef482ace3d7a05e683cc1 for all the changes in 1.10.12

**Pull request merged**